algebird-core/src/main/scala/com/twitter/algebird/maps/redblack.scala

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright 2012 Twitter, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package com.twitter.algebird.maps.redblack
+
+import math.Ordering
+
+object tree {
+  /** The color (red or black) of a node in a Red/Black tree */
+  sealed trait Color extends Serializable
+  case object R extends Color
+  case object B extends Color
+
+  /** Defines the data payload of a tree node */
+  trait Data[K] extends Serializable {
```

Comment:
Covariance involves some open questions for me.  For example, if we want to inherit from SortedMap[K, +V], then I have to support a type-widening insert:
```scala
def +[V2 >: V](kv: (K, V2)): Map[K, V2]
```
Which I think I can do, but it isn't painless either.